File: c++/57_Insert_Interval.cpp
#include <iostream>
#include <vector>
#include <map>
#include <string>
#include <algorithm>

using namespace std;

/**
 * Definition for an interval.
 * struct Interval {
 *     int start;
 *     int end;
 *     Interval() : start(0), end(0) {}
 *     Interval(int s, int e) : start(s), end(e) {}
 * };
 */

class Solution {
public:
    vector<Interval> insert(vector<Interval>& intervals, Interval newInterval) {
        vector<Interval> res;
        res.push_back(newInterval);

        int n = intervals.size();
        if (n == 0) return res;
        int loc = n;
        for(int i = 0; i < n; i++) {
            cout << i << endl;
            if (intervals[i].start > newInterval.start) {
                loc = i;
                break;
            }
        }

        for (int i = loc; i > 0; i--) {
            if (res[0].start <= intervals[i - 1].end) {
                res[0].start = intervals[i - 1].start;
                res[0].end = max(intervals[i - 1].end, res[0].end);
            }
            else
                res.insert(res.begin(), intervals[i - 1]);
        }
        for (int i = loc; i < n; i++) {
            if (res.back().end >= intervals[i].start)
                res.back().end = max(res.back().end, intervals[i].end);
            else
                res.push_back(intervals[i]);
        }

        return res;
    }
};



int main() {
    Solution s;
}

File: python/57_Insert_Interval.py
#!/usr/bin/python
import unittest


class Solution(object):
    def insert(self, intervals, newInterval):
        if len(intervals) == 0:
            return [newInterval]
        for i in range(len(intervals)):
            if intervals[i].start > newInterval.start:
                loc = i # binary search would be better, just skip it
                break
        if intervals[-1].start <= newInterval.start: loc = i + 1
        res = [newInterval]
        for i in range(loc, 0,  -1):
            if intervals[i - 1].end >= res[0].start:
                res[0].start = intervals[i - 1].start
                res[0].end = max(res[0].end, intervals[i - 1].end)
            else:
                res.insert(0, intervals[i - 1])

        for i in range(loc, len(intervals)):
            if res[-1].end >= intervals[i].start:
                res[-1].end = max(intervals[i].end, res[-1].end)
            else:
                res.append(intervals[i])
        return res
                


        

def main():
    s = Solution()
    s.insert()


if __name__ == "__main__":
    main()
